saleor/discount/utils.py

```diff
@@ -376,17 +376,31 @@ def create_checkout_line_discount_objects_for_catalogue_promotions(
     ) = discount_data
 
     new_line_discounts = []
-    if discounts_to_create_inputs:
-        new_line_discounts = [
-            CheckoutLineDiscount(**input) for input in discounts_to_create_inputs
-        ]
-        CheckoutLineDiscount.objects.bulk_create(new_line_discounts)
+    with transaction.atomic():
+        # Protect against potential thread race. CheckoutLine object can have only
+        # single catalogue discount applied.
+        line_ids = [line_info.line.id for line_info in lines_info]
+        _lines_lock = list(
+            CheckoutLine.objects.filter(id__in=line_ids).select_for_update(
```

Comment:
I belive the reason we had a deadlock was: https://github.com/saleor/saleor/pull/15599#discussion_r1535712793. Previously, instead of update, we were making add&delete.

Reply:
I locked whole checkout instead of lines.